File: Assignment_2/a2_starter/Task4.scala
import org.apache.spark.{SparkContext, SparkConf}
import scala.math.min
import scala.collection.Map
import org.apache.spark.rdd.RDD

// Performance (with broadcast and hashmap)
// 45.53user 5.19system 1:09.62elapsed 72%CPU (0avgtext+0avgdata 773204maxresident)k
// 0inputs+16outputs (0major+377086minor)pagefaults 0swaps
// Found 3 items
// -rw-r--r--   2 z498zhan z498zhan          0 2019-06-29 15:50 /user/z498zhan/a2_starter_code_output/_SUCCESS
// -rw-r--r--   2 z498zhan z498zhan    1254855 2019-06-29 15:50 /user/z498zhan/a2_starter_code_output/part-00000
// -rw-r--r--   2 z498zhan z498zhan    1344210 2019-06-29 15:50 /user/z498zhan/a2_starter_code_output/part-00001
// 1254855 + 1344210 = 2599065

// Old Performance (with built-in cartesian product)
// 49.17user 5.43system 7:05.98elapsed 12%CPU (0avgtext+0avgdata 754144maxresident)k
// 0inputs+16outputs (0major+381832minor)pagefaults 0swaps
// Found 5 items
// -rw-r--r--   2 z498zhan z498zhan          0 2019-06-29 16:04 /user/z498zhan/a2_starter_code_output/_SUCCESS
// -rw-r--r--   2 z498zhan z498zhan     652680 2019-06-29 16:00 /user/z498zhan/a2_starter_code_output/part-00000
// -rw-r--r--   2 z498zhan z498zhan     602175 2019-06-29 16:00 /user/z498zhan/a2_starter_code_output/part-00001
// -rw-r--r--   2 z498zhan z498zhan     703185 2019-06-29 16:04 /user/z498zhan/a2_starter_code_output/part-00002
// -rw-r--r--   2 z498zhan z498zhan     641025 2019-06-29 16:03 /user/z498zhan/a2_starter_code_output/part-00003
// 652680 + 602175 + 703185 + 641025 = 2599065

object Task4 {
  def findSimilarity(ratings1: Array[Byte], ratings2: Array[Byte]): Int = {
    var similarity = 0

    for(i <- 0 until ratings1.length){
      if (ratings1(i) == ratings2(i) && ratings1(i) != 0) {
        similarity += 1
      }
    }

    return similarity
  } 

  def buildMovieRatingsMap(movieRatings: RDD[String]): Map[String, Array[Byte]] = {
    movieRatings.map(movieRating => {
      val tokens = movieRating.split(",", -1)
      val title = tokens(0)
      val ratings = new Array[Byte](tokens.length - 1)

      for (i <- 0 until ratings.length) {
        if (tokens(i + 1) != "") {
          ratings(i) = tokens(i + 1).toByte
        }
      }
      (title, ratings)
    }).collectAsMap()
  }

  def main(args: Array[String]) {
    val conf = new SparkConf().setAppName("Task 4")
    val sc = new SparkContext(conf)

    val textFile = sc.textFile(args(0))

    val movieRatingsMap = sc.broadcast(buildMovieRatingsMap(textFile))

    val output = textFile.flatMap(movie1 => {
      val title1 = movie1.split(",", 2)(0)
      val ratings1 = movieRatingsMap.value(title1)

      movieRatingsMap.value
        .filterKeys(title2 => title2 > title1)
        .map(movie2 => {
          val title2 = movie2._1
          val ratings2 = movie2._2

          val similarity = findSimilarity(ratings1, ratings2)
          (title1 + "," + title2 + "," + similarity)
        })
    }).saveAsTextFile(args(1))
  }
}

File: Assignment_2/a2_starter/Task4.java
import java.io.IOException;
import java.util.HashMap;
import java.util.StringTokenizer;

import org.apache.hadoop.conf.Configuration;
import org.apache.hadoop.fs.Path;
import org.apache.hadoop.io.IntWritable;
import org.apache.hadoop.io.NullWritable;
import org.apache.hadoop.io.Text;
import org.apache.hadoop.mapreduce.Job;
import org.apache.hadoop.mapreduce.Mapper;
import org.apache.hadoop.mapreduce.Reducer;
import org.apache.hadoop.mapreduce.lib.input.TextInputFormat;
import org.apache.hadoop.mapreduce.lib.output.TextOutputFormat;
import org.apache.hadoop.util.GenericOptionsParser;
import org.apache.hadoop.filecache.DistributedCache;
import java.net.URI;
import java.io.BufferedReader;
import java.io.FileReader;
import java.io.*;  

// Performance: 11.59user 0.92system 0:49.22elapsed 25%CPU (0avgtext+0avgdata 401988maxresident)k
public class Task4 {

  public static class MovieSimilarityMapper extends Mapper<Object, Text, Text, NullWritable> {
    private Text result = new Text();
    private HashMap<String, Byte[]> movieRatingsMap = new HashMap<String, Byte[]>();

    public void setup(Context context) throws IOException, InterruptedException {
      Path cachePath = context.getLocalCacheFiles()[0];
      BufferedReader reader = new BufferedReader(new FileReader(cachePath.toString())); 
      String line;
      
      while ((line = reader.readLine()) != null){
        String[] tokens = line.split(",", -1);
        String title = tokens[0];
        Byte[] ratings = new Byte[tokens.length - 1];

        for (int i = 0; i < ratings.length; i++) {
          if (!tokens[i + 1].isEmpty()) {
            ratings[i] = Byte.parseByte(tokens[i + 1]);
          } else {
            ratings[i] = 0;
          }
        }

        movieRatingsMap.put(title, ratings);
      }
    }

    public void map(Object key, Text value, Context context) throws IOException, InterruptedException {
      String title1 = value.toString().split(",", 2)[0];
      Byte[] ratings1 = movieRatingsMap.get(title1);

      for (String title2 : movieRatingsMap.keySet()) {
        if (title2.compareTo(title1) > 0) {
          Byte[] ratings2 = movieRatingsMap.get(title2);
          int similarity = findSimilarity(ratings1, ratings2);

          result.set(title1 + "," + title2 + "," + similarity);
          context.write(result, NullWritable.get());
        }
      }
    }

    public int findSimilarity(Byte[] ratings1, Byte[] ratings2) {
      int similarity = 0;

      for (int i = 0; i < Math.min(ratings1.length, ratings2.length); i++) {
        if (ratings1[i] == ratings2[i] && ratings1[i] != 0) {
          similarity ++;
        }
      }

      return similarity;
    }
  }

  public static void main(String[] args) throws Exception {
    Configuration conf = new Configuration();
    conf.set("mapred.textoutputformat.separator", ",");

    String[] otherArgs = new GenericOptionsParser(conf, args).getRemainingArgs();

    Job job = new Job(conf, "Task4");
    job.setJarByClass(Task4.class);

    job.setMapperClass(MovieSimilarityMapper.class);
    job.setOutputKeyClass(Text.class);
    job.setOutputValueClass(NullWritable.class);

    // Set up distributed cache
    job.addCacheFile(new URI(otherArgs[0]));
    // This is a mapper-only hadoop job
    job.setNumReduceTasks(0);

    TextInputFormat.addInputPath(job, new Path(otherArgs[0]));
    TextOutputFormat.setOutputPath(job, new Path(otherArgs[1]));

    System.exit(job.waitForCompletion(true) ? 0 : 1);
  }
}
